### PR TITLE
Interface with Tunix

### PR DESCRIPTION
**Pull request by A9isha**
# Description

Add a wrapper around MaxText's Transformer (which is NNX module) class to be compatible with Tunix's call function. Also, register `HyperParameters` as `pytree_node_class`. This was needed during checkpoint loading such that Jax doesn't traverse inside the `HyperParameters` class

# Tests

Tested by running 

`python3 grpo_llama3_demo1.py`

on the GRPO script https://github.com/google/tunix/blob/anisha-tunix-grpo-kaggle/examples/grpo_llama3_demo1.py. And compared the outputs before and after the change.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import jax
+import jax.numpy as jnp
+from flax import nnx
+
+from MaxText.common_types import MODEL_MODE_TRAIN, MODEL_MODE_AUTOREGRESSIVE, Config
+from MaxText.layers.models import Transformer
+
+Array = jax.Array  # Alias for clarity
+
+class TunixMaxTextLlama(nnx.Module):
```
Comment (gagika) [inline MaxText/integration/tunix/tunix_adaptor.py]: is this Llama specific?

If not, please name it to more generic name, e.g. TunixMaxTextAdapter

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        input_tokens: Array,             # [B, L]
+        positions: Array,                # [B, L]
+        cache: Optional[Any],            # Tunix passes None in SFT
+        attention_mask: Optional[Array], # [B, L, L] or None
+        output_hidden_states: bool = False,  # ignored
+    ) -> Tuple[Array, None]:
+        """Forward compatible with Tunix PeftTrainer default loss.
+        Returns logits, None (cache unused in SFT).
+        """
+        attn = attention_mask if self.use_attention_mask else None
```
Comment (gagika) [inline MaxText/integration/tunix/tunix_adaptor.py]: how is attn used?

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+from __future__ import annotations
+
+from typing import Optional, Tuple, Any
+
+import jax
+import jax.numpy as jnp
+from flax import nnx
+
+from MaxText.common_types import MODEL_MODE_TRAIN, MODEL_MODE_AUTOREGRESSIVE, Config
```
Comment (gagika) [inline MaxText/integration/tunix/tunix_adaptor.py]: are those used? if not please remove.

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    # ------------------------------------------------------------------ #
+    # Tunix call signature (SFT)
+    # ------------------------------------------------------------------ #
+    def __call__(
+        self,
+        input_tokens: Array,             # [B, L]
+        positions: Array,                # [B, L]
+        cache: Optional[Any],            # Tunix passes None in SFT
+        attention_mask: Optional[Array], # [B, L, L] or None
```
Comment (gagika) [inline MaxText/integration/tunix/tunix_adaptor.py]: how does it integrate with Flash attention and lazy attention masking?

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+from __future__ import annotations
```
Comment (gagika) [inline MaxText/integration/tunix/tunix_adaptor.py]: could you use two spaces for indentation (use `bash code_style.sh`) to follow MaxText style.

Any plans for tests for this adapter?

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+from typing import Optional, Tuple, Any
+
+import jax
+import jax.numpy as jnp
+from flax import nnx
+
+from MaxText.common_types import MODEL_MODE_TRAIN, MODEL_MODE_AUTOREGRESSIVE, Config
+from MaxText.layers.models import Transformer
+
+Array = jax.Array  # Alias for clarity
```
Comment (richjames0) [inline MaxText/integration/tunix/tunix_adaptor.py]: from jax import Array is probably more idiomatic?

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        positions: Array,                # [B, L]
+        cache: Optional[Any],            # Tunix passes None in SFT
+        attention_mask: Optional[Array], # [B, L, L] or None
+        output_hidden_states: bool = False,  # ignored
+    ) -> Tuple[Array, None]:
+        """Forward compatible with Tunix PeftTrainer default loss.
+        Returns logits, None (cache unused in SFT).
+        """
+        attn = attention_mask if self.use_attention_mask else None
+        # logits = self.base.forward_train(
```
Comment (richjames0) [inline MaxText/integration/tunix/tunix_adaptor.py]: dead code?

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 2 above)
Comment (richjames0) [reply MaxText/integration/tunix/tunix_adaptor.py]: yeah it looks unused

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 3 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: done

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 6 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: done

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 1 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: done

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 4 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: discussed offline. MaxText creates own attention mask

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 2 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: Apologies. Removed it now.

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 7 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: removed now

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 5 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: updated spaces.

Discussed offline about plans of further tests

File: MaxText/integration/tunix/tunix_adaptor.py
Comment (gagika) [inline MaxText/integration/tunix/tunix_adaptor.py:1]: nit: tunix_adaptor.py → tunix_adapter.py

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    logits = self.base(
+        decoder_input_tokens=input_tokens,
+        decoder_positions=positions,
+        # TODO: @mazumdera - add support for packing
+        decoder_segment_ids=None,
+    )
+    return logits, None
+
+
+def create_maxtext_to_vllm_mappings():
```
Comment (gagika) [inline MaxText/integration/tunix/tunix_adaptor.py]: this mapping is Llama2/3 specific, let's rename it to reflect that.

Also any reason to keep this in the tunix_adapter file instead of something like this: MaxText/integration/vllm/llama_param_mapping.py

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+      "base.decoder.logits_dense.kernel": (
+          "model.lm_head",
+          (None, "model"),
+      ),
+      # Layer-specific mappings (scanned -> unscanned)
+      # MLP components - shard hidden dimensions for TP
+      "base.decoder.layers.mlp.wi_0.kernel": (
+          "model.layers.*.mlp.gate_proj.kernel",
+          (None, "layer", "model"),
+      ),  # gate_proj: (4096, 14336) - shard output
```
Comment (gagika) [inline MaxText/integration/tunix/tunix_adaptor.py]: are all this comments (here and below) needed?

The shape (4096, 14336) might be true for Llama3-8b but not Llama3-70b. Let's use something like this instead: (base_emb_dim, base_mlp_dim)

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+from typing import Optional, Tuple, Any
+
+from jax import Array
+from flax import nnx
+from MaxText.layers.models import Transformer
+
+
+
+
+class TunixMaxTextAdaptor(nnx.Module):
```
Comment (gagika) [inline MaxText/integration/tunix/tunix_adaptor.py]: TunixMaxTextAdapter ?

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        decoder_input_tokens=input_tokens,
+        decoder_positions=positions,
+        # TODO: @mazumdera - add support for packing
+        decoder_segment_ids=None,
+    )
+    return logits, None
+
+
+def create_maxtext_to_vllm_mappings():
+  """Create mappings for transferring MaxText scanned state to vLLM unscanned state."""
```
Comment (gagika) [inline MaxText/integration/tunix/tunix_adaptor.py]: can you add comments on what are sharding expectations on MaxText and vLLM?

it would be nice if sharding is handled in separate thinner adapter/functionality, as it's quite independant of weight mapping. 

cc
@wenxindongwork

**Review by richjames0 — APPROVED** [review]
lgtm when gagik's happy :)

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 11 above)
Comment (wenxindongwork) [reply MaxText/integration/tunix/tunix_adaptor.py]: Currently the param mapping conforms to the Tunix API, which combines the param name and sharding in one dictionary.

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 8 above)
Comment (wenxindongwork) [reply MaxText/integration/tunix/tunix_adaptor.py]: thanks for the suggestion! let me move the mapping to a separate file

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 11 above)
Comment (gagika) [reply MaxText/integration/tunix/tunix_adaptor.py]: are there plans to decouple them?

@annyan09023

Comment (annyan09023) [reply MaxText/integration/tunix/tunix_adaptor.py]: I think @wang2yn84 decided most of API, when he is back, he can decide whether to decouple and how to do it properly.

File: MaxText/integration/tunix/tunix_adaptor.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
+from __future__ import annotations
+
+from typing import Optional, Tuple, Any
+
+from jax import Array
+from flax import nnx
+from MaxText.layers.models import Transformer
+
+
```
Comment (SurbhiJainUSC) [inline MaxText/integration/tunix/tunix_adaptor.py]: nit: multiple lines

File: MaxText/integration/tunix/tunix_adaptor.py
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py:1]: done

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 12 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: made it into 1

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 8 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: we now have MaxText/integration/tunix/weight_mapping/llama3.py - thanks WenXin!

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 10 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: done

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 11 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: done, added a comment 

```
  def to_hf_mapping():
    """
    Mapping from MaxText model to HuggingFace vLLM model.
    Currently the param mapping conforms to the Tunix API, which combines the param name and sharding in one dictionary.
    This is subject to change in the future where we can decouple the two.
    """
```

File: MaxText/integration/tunix/weight_mapping/llama3.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+    def transform_query_kernel(arr):
+      head_dim = arr.shape[-1]
+      depth_scale = np.dtype("float32").type(np.sqrt(head_dim))
+      arr = arr * depth_scale
+      return reorder_rope(arr)
+
+    hook_fns = {
+        "base.decoder.layers.self_attention.query.kernel": transform_query_kernel,
+        "base.decoder.layers.self_attention.key.kernel": reorder_rope,
```
Comment (gagika) [inline MaxText/integration/tunix/weight_mapping/llama3.py:39]: @wenxindongwork I recall that this is only needed for Llama2/Llama3 not Llama3.1/Llama3 (or the other way around), perhaps we need to unify it for MaxText.

@A9isha 
Maybe this issue was causing convergence issues?

**Review by gagika — APPROVED** [review]
LGTM, Please double check when q/k proj need for RoPE permutation as it's likely different for Llama2,3 and Llama 3.1 (Llama4).

File: MaxText/integration/tunix/tunix_adaptor.py (hunk 9 above)
Comment (A9isha) [reply MaxText/integration/tunix/tunix_adaptor.py]: removed it now.

File: MaxText/integration/tunix/weight_mapping/llama3.py (hunk 13 above)
Comment (A9isha) [reply MaxText/integration/tunix/weight_mapping/llama3.py:39]: Gagik, WenXin and I discussed offline and we are specifying and we support for Llama3.1 only at the moment. Follow up PR to support more models.
